[FrameworkBundle][Translation][TwigBridge] Add option to extract blank messages

gndk opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | 6.2
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Tickets       | -
| License       | MIT
| Doc PR        | symfony/symfony-docs#16621

Currently, new messages are extracted as a prefixed copy of the key (`"foo" -> "__foo"`) when using the `translation:extract` command. 

With his PR I propose a `--blank` option that would extract new messages as blank strings (`"foo" -> ""`). 

This allows for smoother workflows with translation providers like Loco, Lokalise or CrowdIn, because by some of them the prefixed copy is considered **translated**, while blank strings are considered **untranslated**. 

The result is that you currently have 100% translation progress immediately after `translation:push`, even if *none* of the translation keys are actually translated. This makes working with hundreds or thousands of translation keys or general collaboration through these tools unnecessarily hard. 

For example lets say you have 1000 translation keys in your application and want to add a new locale. You extract and push them to the provider. Then you or some translator starts work on them. They won't finish all keys in one session, but instead maybe finish 100 of them and continue work the next day. However, because all the messages already have the "translated" status, its hard to know where to continue. You have to look at every single message again and determine if it is **really** translated, or **if the strings are just technically different**. 

**It is very easy to miss untranslated messages if all of them have a green checkmark, regardless of the real translation status.**

The benefit of pushing blank strings as translation messages is that it allows you to **easily filter for untranslated messages as a "todo list"**, instead of checking each message manually for its translation status.

gndk commented:
Psalm fail seems unrelated as I see the same error in other PRs. AppVeyor timeout.

gndk commented:
Pinging @Nyholm, because you agreed in my previous PR that this is a problem worth solving 🙂 .

ro0NL commented on src/Symfony/Component/Translation/Extractor/BlankableExtractorInterface.php:
```diff
@@ -0,0 +1,41 @@
+    /**
+     * Extract new-found messages as blank strings.
+     */
+    public function blank(bool $blank);
```
i tend to prefer a new virtual method on existing ExtractorInterface (`@method` annotation)

gndk replied:
Using the existing interface would make the change much smaller and easier. I didn't know that was possible from my understanding of the BC guidelines.

So your suggestion would basically be this and removing the new `BlankableExtractorInterface`? (plus the resulting follow-up changes of course)


```
diff --git a/src/Symfony/Component/Translation/Extractor/ExtractorInterface.php b/src/Symfony/Component/Translation/Extractor/ExtractorInterface.php
index b76a7f20ba..e8745c6d15 100644
--- a/src/Symfony/Component/Translation/Extractor/ExtractorInterface.php
+++ b/src/Symfony/Component/Translation/Extractor/ExtractorInterface.php
@@ -18,6 +18,8 @@ use Symfony\Component\Translation\MessageCatalogue;
  * New found messages are injected to the catalogue using the prefix.
  *
  * @author Michel Salib <michelsalib@hotmail.com>
+ *
+ * @method blank(bool $blank)
  */
 interface ExtractorInterface
 {

```

ro0NL replied:
for this design it's the fastest route yes (i think it's sensible design :+1: )

gndk replied:
I made the changes in a separate commit. Like this approach much better 👍

ro0NL commented on src/Symfony/Bundle/FrameworkBundle/Command/TranslationUpdateCommand.php:
```diff
@@ -391,10 +396,13 @@ private function filterCatalogue(MessageCatalogue $catalogue, string $domain): M
         $this->extractor->setPrefix($prefix);
+        if (method_exists($this->extractor, 'blank')) {
+            $this->extractor->blank($blank);
+        }
```
`elseif($blank) { throw useful message; }` would help for sf/translation@5.4, or implementations that aren't upgraded yet

gndk replied:
Done 👍

welcoMattic commented on src/Symfony/Bridge/Twig/CHANGELOG.md:
```diff
@@ -1,6 +1,10 @@
 
+6.1
+---
+* Add the option to extract blank strings instead of a prefixed copy of the translation key with the new `blank()' method in the `TwigExtractor`
```
```suggestion
* Add the option to extract blank strings instead of a prefixed copy of the translation key with the new `blank()` method in the `TwigExtractor`
```

welcoMattic commented:
@gndk great work, thank you! Could you please rebase your branch and address the 2 quick comments? Thanks :+1:

gndk committed:
[FrameworkBundle][Translation][TwigBridge] Add option to extract blank messages

Extract new-found messages as blank strings instead of
a prefixed copy of the translation key when using
the blank option for the translation:extract command.

This allows for a smoother workflow when using third
party translation providers, because empty strings
are considered “untranslated”, while an exact copy of
the translation key is considered “translated”.

gndk commented on src/Symfony/Bundle/FrameworkBundle/Command/TranslationUpdateCommand.php:
```diff
@@ -148,6 +153,12 @@ protected function execute(InputInterface $input, OutputInterface $output): int
             return 1;
         }
 
+        if (! method_exists($this->extractor, 'blank') && true === $input->getOption('blank')) {
```
I added the error message up here because other checks like this also live here. And I don't need to pass the OutputInterface down to the method. Seems cleaner this way.

Not sure about the wording of the message. Anything you want to change or add?

gndk commented on src/Symfony/Bundle/FrameworkBundle/Tests/Command/TranslationUpdateCommandTest.php:
```diff
@@ -37,6 +37,17 @@ public function testDumpMessagesAndCleanWithDeprecatedCommandName()
+    public function testDumpBlankMessagesAndCleanWithDeprecatedCommandName()
+    {
+        $tester = $this->createCommandTester(['messages' => ['foo' => '']]);
+        if (!method_exists($tester, 'blank')) {
+            $this->markTestSkipped('The --blank option is not supported by this extractor');
+        }
```
Not sure if thats the best approach to handle failing tests in low deps jobs, but I think its reasonable :)

gndk replied:
Seems like they are always skipped now... That was not intended.

gndk replied:
It seems to work as intended in CI. Tests are run in the `8.1, high-deps` job, but skipped in `8.1, low-deps`.

gndk replied:
Thinking about it, it would actually be better to assert the output based on that condition. So normal output if method exists and error message if not. Will change that.

gndk commented on src/Symfony/Bundle/FrameworkBundle/Command/TranslationUpdateCommand.php:
```diff
@@ -148,6 +153,12 @@ protected function execute(InputInterface $input, OutputInterface $output): int
             return 1;
         }
 
+        if (!method_exists($this->extractor, 'blank') && true === $input->getOption('blank')) {
+            $errorIo->error('The --blank option is not supported by this extractor. It is available in symfony/translation 6.1 and symfony/twig-bridge 6.1.');
+
+            return 1;
+        }
+
```
I added the error message up here because other checks like this also live here. And I don't need to pass the OutputInterface down to the method. Seems cleaner this way.

Not sure about the wording of the message. Anything you want to change or add?

welcoMattic commented on src/Symfony/Bridge/Twig/CHANGELOG.md:
```diff
@@ -5,6 +5,8 @@ CHANGELOG
 ---
 
  * Wrap help messages on form elements in `div` instead of `p`
+ * Implement the new `BlankableExtractorInterface` in the `TwigExtractor`, which allows extracting blank strings instead of a prefixed copy of the translation key
```
This interface doesn't exist, right?

gndk replied:
Oops, good catch! Must have messed that up in the rebase. Fixed.

nicolas-grekas reviewed (changes requested):
The proposed method is a setter. This makes the extractor services mutable. We'd really better avoid that.
Does it really make sense to have a command line option for this? Isn't setting messages as empty a project-wide setting? If yes, then we'd better go with a config option and a constructor argument.

welcoMattic commented:
@nicolas-grekas is right, we should avoid mutable services, I missed it. :+1: for config option.

gndk commented:
Ok, I will turn it into a config option instead 👍 . Don't have time this week, so will be done after the easter weekend.

nicolas-grekas commented:
Friendly ping @gndk
